src/types/typeobject.rs

```diff
@@ -30,9 +33,52 @@ impl PyType {
         py.from_borrowed_ptr(p as *mut ffi::PyObject)
     }
 
-    /// Gets the name of the `PyType`.
-    pub fn name(&self) -> PyResult<&str> {
-        self.getattr(intern!(self.py(), "__qualname__"))?.extract()
+    /// Gets the [qualified name](https://docs.python.org/3/glossary.html#term-qualified-name) of the `PyType`.
+    pub fn qualname(&self) -> PyResult<String> {
+        #[cfg(any(Py_LIMITED_API, PyPy, not(Py_3_11)))]
+        let name = self.getattr(intern!(self.py(), "__qualname__"))?.extract();
+
+        #[cfg(not(any(Py_LIMITED_API, PyPy, not(Py_3_11))))]
+        let name = {
+            use crate::ffi_ptr_ext::FfiPtrExt;
+            use crate::types::any::PyAnyMethods;
+
+            let obj = unsafe {
+                ffi::PyType_GetQualName(self.as_type_ptr()).assume_owned_or_err(self.py())?
+            };
+
+            obj.extract()
+        };
+
+        name
+    }
+
+    /// Gets the full name, which includes the module, of the `PyType`.
+    pub fn name(&self) -> PyResult<Cow<'_, str>> {
+        #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+        {
+            let name = unsafe { CStr::from_ptr((*self.as_type_ptr()).tp_name) }.to_str()?;
```

Comment:
Now I stare at this, I wonder if we need to be defensive against reassignments to `tp_name`? Looking at CPython source, it looks like this is allowed if the type doesn't have flag `Py_TPFLAGS_IMMUTABLETYPE`.

Reply:
So copy out the value if the flag is not set?

Reply:
This flags seems to exist only for 3.10+, so always copy out before that version?

Reply:
Yeah, I suspect that's probably the correct approach 👍